packages/compiler/babel.ts

```diff
@@ -0,0 +1,166 @@
+import * as t from '@babel/types';
+import { addNamed } from '@babel/helper-module-imports';
+import { resolveImportSource } from './utils/mod';
+import { transformFor } from './for';
+import { catchError } from './utils/log';
+import type { Options } from './options';
+import type { NodePath, PluginObj } from '@babel/core';
+import { transformBlock } from './block';
+import { parseAuto } from './auto';
+
+export interface Info {
+  imports: {
+    block: string | null;
+    For: string | null;
+    renderReactScope: string | null;
+    aliases: {
+      block: Set<string> | null;
+      For: Set<string> | null;
+      renderReactScope: Set<string> | null;
+    };
+    source: string | null;
+  };
+  programPath: NodePath<t.Program>;
+  blocks: Map<string, t.Identifier>;
+  filename: string;
+  options: Options;
+  // TODO: add debug that stores timings
+}
+
+export function babel(
+  _: unknown,
+  options: Options,
+  filename: string
+): PluginObj {
+  // type State = SetAllNullable<Info, 'aliases' | 'blocks'>;
+  const state: Info['imports'] = {
+    block: null,
+    For: null,
+    renderReactScope: null,
+    aliases: {
+      block: null,
+      For: null,
+      renderReactScope: null,
+    },
+    source: null,
+  };
+  let info: Info | null = null;
+
+  return {
+    name: 'million',
+    visitor: {
+      Program: {
+        enter(programPath: NodePath<t.Program>) {
+          info = {
+            imports: state,
+            programPath,
+            blocks: new Map<string, t.Identifier>(),
+            filename,
+            options,
+          };
+
+          programPath.traverse<Info['imports']>(
+            {
+              ImportDeclaration(path, state) {
+                const importDeclaration = path.node;
+                const importSource = importDeclaration.source;
+                if (!importSource.value.includes('million')) return;
```

Comment:
Wait how accurate is this or probably the other question is, what patterns are you trying to check against